File: python/dsc/context.py
# Copyright (c) 2024-2025, Christian Gilli <christian.gilli@dspcraft.com>
# All rights reserved.
#
# This code is licensed under the terms of the 3-clause BSD license
# (https://opensource.org/license/bsd-3-clause).

from ._bindings import _dsc_ctx_init, _dsc_print_mem_usage, _dsc_set_default_device
from .device import _get_device, DeviceType
import psutil

_ctx_instance = None


class _DscContext:
    def __init__(self, main_mem: int):
        self._ctx = _dsc_ctx_init(main_mem)

    # TODO: (3)
    # def __del__(self):
        # _dsc_ctx_free(self._ctx)


def _get_ctx():
    global _ctx_instance
    if _ctx_instance is None:
        # Workaround: instead of throwing an error if the context is not initialized
        # we can simply initialize one with a fixed amount of memory that is a small %
        # of the total available memory.
        total_mem = psutil.virtual_memory().total
        mem = int(total_mem * 0.1)
        print(
            f'DSC has not been explicitly initialized. Using {round(mem / (1024. * 1024.))}MB.'
            f' If you require more memory please call dsc.init() once before executing your code.'
        )
        _ctx_instance = _DscContext(mem)
    return _ctx_instance._ctx


def init(mem_size: int):
    global _ctx_instance
    if _ctx_instance is None:
        _ctx_instance = _DscContext(mem_size)
    else:
        raise RuntimeWarning('Context already initialized')

def print_mem_usage():
    _dsc_print_mem_usage(_get_ctx())

def set_default_device(device: DeviceType):
    _dsc_set_default_device(_get_ctx(), _get_device(device))

File: python/src/context.cpp
// Copyright (c) 2024-2025, Christian Gilli <christian.gilli11@gmail.com>
// All rights reserved.
//
// This code is licensed under the terms of the 3-clause BSD license
// (https://opensource.org/license/bsd-3-clause).

#include <nanobind/nanobind.h>
// To convert to and from Python str
#include <nanobind/stl/string.h>
#include <sys/sysinfo.h>
#include <unordered_map>
#include "utils.h"
#include "dsc.h"


namespace nb = nanobind;
using namespace nb::literals;


// TODO: note that if there are concurrency issues this is probably the root cause (should be atomic!)

static dsc_ctx *g_ctx = nullptr;

namespace dsc::ctx {
void init(const usize mem_size) noexcept {
    g_ctx = dsc_ctx_init(mem_size);
}

void teardown() noexcept {
    if (g_ctx) dsc_ctx_free(g_ctx);
    g_ctx = nullptr;
}

dsc_ctx *get() noexcept {
    if (!g_ctx) {
        struct sysinfo info;
        DSC_ASSERT(sysinfo(&info) == 0);
        const usize total_ram = info.totalram * info.mem_unit;
        printf("DSC has not been explicitly initialized so it will try to reserve 80%% of the total available memory in the system.\n"
               "To explicitly initialize DSC call `dsc.init()` with a specific memory size in bytes.\n");
        init((usize) total_ram * 0.8);
    }
    return g_ctx;
}
}


void init_context(nb::module_& m) {
    nb::enum_<dsc_device_type>(m, "Device")
        .value("default", DEFAULT)
        .value("cpu", CPU)
        .value("gpu", GPU)
        .export_values();

    m.def("init", [](const usize mem_size) {
        dsc::ctx::init(mem_size);
    }, "nb"_a);

    // m.attr("_finalizer") = nb::capsule((void *) 1, "dsc_ctx", [](void *) noexcept {
    //    dsc::ctx::teardown();
    // });

    m.def("print_mem_usage", []() {
        dsc_print_mem_usage(dsc::ctx::get());
    });

    m.def("set_default_device", [](const Device &device) noexcept {
       dsc_set_default_device(dsc::ctx::get(), get_dsc_device(device));
    }, "device"_a);
}
